torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -632,10 +645,9 @@ void ProcessGroupNCCL::workCleanupLoop() {
     for (auto it = workList_.begin(); it != workList_.end();
          /* no increment*/) {
       auto& work = *it;
-      if (work->isCompleted()) {
-        // Handle Exceptions on failed GPU operations and remove completed
-        // workNCCL objects from work vector.
-        work->handleNCCLGuard();
+      // Handle Exceptions on failed GPU operations and remove completed
+      // workNCCL objects from work vector.
+      if (work->isCompletedAndThrowException()) {
```

Comment:
It would be nice to add a LOG(ERROR) before we throw an exception mentioning the following: "Some NCCL operations have timed out/failed and due to the async nature of CUDA kernels subsequent GPU operations might run on corrupted/incomplete data. To avoid this inconsistency, we're taking the entire process down."

Reply:
Created https://github.com/pytorch/pytorch/pull/44988 with this change